pkg/kwok/server/debugging_port_forword.go

```diff
@@ -20,20 +20,66 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"net"
 	"net/http"
 
 	"github.com/emicklei/go-restful/v3"
 	"k8s.io/apimachinery/pkg/types"
 
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
 	"sigs.k8s.io/kwok/pkg/kwok/server/portforward"
 	"sigs.k8s.io/kwok/pkg/log"
+	"sigs.k8s.io/kwok/pkg/utils/exec"
+	"sigs.k8s.io/kwok/pkg/utils/slices"
 )
 
 // PortForward handles a port forwarding request.
 func (s *Server) PortForward(ctx context.Context, podName, podNamespace string, uid types.UID, port int32, stream io.ReadWriteCloser) error {
-	// TODO: Configure and implement the port forward streamer
-	msg := fmt.Sprintf("TODO: PortForward(%q, %q)", podName+"/"+podNamespace, port)
-	_, _ = stream.Write([]byte(msg))
+	defer func() {
+		_ = stream.Close()
+	}()
+
+	forward, err := s.getPodsForward(podName, podNamespace, port)
+	if err != nil {
+		return err
+	}
+
+	if len(forward.Command) > 0 {
+		cmdStream := exec.IOStreams{
+			In:  stream,
+			Out: stream,
+		}
+		return exec.Exec(ctx, "", cmdStream, forward.Command[0], forward.Command[1:]...)
+	}
+
+	addr := fmt.Sprintf("%s:%d", forward.TargetAddress, forward.TargetPort)
+	dial, err := net.Dial("tcp", addr)
+	if err != nil {
+		return fmt.Errorf("failed to dial %s: %w", addr, err)
+	}
+	defer func() {
+		_ = dial.Close()
+	}()
+
+	done := make(chan struct{})
+	var (
+		dstErr error
+		srcErr error
+	)
+	go func() {
+		_, err = io.Copy(dial, stream)
+		close(done)
+	}()
+
+	_, srcErr = io.Copy(stream, dial)
+	<-done
+
+	if srcErr != nil {
+		return srcErr
+	}
+	if dstErr != nil {
+		return dstErr
+	}
 	return nil
 }
 
```

Comment:
Wrap the creation of a stream in a context-responsive function

```suggestion

// tunnel create tunnels for two streams.
func tunnel(ctx context.Context, c1, c2 io.ReadWriter, buf1, buf2 []byte) error {
	errCh := make(chan error)
	go func() {
		_, err := io.CopyBuffer(c2, c1, buf1)
		errCh <- err
	}()
	go func() {
		_, err := io.CopyBuffer(c1, c2, buf2)
		errCh <- err
	}()
	select {
	case <-ctx.Done():
		// Do nothing
	case err1 := <-errCh:
		select {
		case <-ctx.Done():
			if err1 != nil {
				return err1
			}
			// Do nothing
		case err2 := <-errCh:
			if err1 != nil {
				return err1
			}
			return err2
		}
	}

	if err := ctx.Err(); err != nil && !errors.Is(err, context.Canceled) {
		return err
	}
	return nil
}
```

Reply:
What size of the buffer would make sense here? The kubelet's lifecycle/handlers.go specifies `maxRespBodyLength` of 10 KB. I don't know what size of the buffer should be. I guess `io.Copy` should not be used to control allocations, right?

Reply:
https://github.com/golang/go/blob/01a5a83cfbe594538f6fe9f49bd27d9dc36155a4/src/io/io.go#L416

I think using 32*1024 would be appropriate


https://github.com/golang/go/blob/def0be5e34d6657c84407c8afe4ace05d10d1848/src/io/io.go#L423

And this can be removed in go1.21, we leave a TODO comment.